tests/test_api_status.py

```diff
@@ -0,0 +1,113 @@
+                    WHERE type != 'transaction'
+                    AND project_id = {self.project_id}
+                    AND timestamp >= toDateTime('{self.base_time.isoformat()}')
+                    AND timestamp < toDateTime('{self.next_time.isoformat()}')""",
+                    "turbo": False,
+                    "consistent": True,
+                    "debug": True,
+                }
+            ),
+            headers={"referer": "test"},
+        )
+
+    def setup_method(self, test_method: Callable[..., Any]) -> None:
+        super().setup_method(test_method)
+        self.event = get_raw_event()
+        self.project_id = self.event["project_id"]
+        self.base_time = datetime.utcnow().replace(
+            minute=0, second=0, microsecond=0
+        ) - timedelta(minutes=180)
+        self.next_time = datetime.utcnow().replace(
+            minute=0, second=0, microsecond=0
+        ) + timedelta(minutes=180)
+
+    @patch("snuba.settings.RECORD_QUERIES", True)
+    @patch("snuba.state.record_query")
+    @patch("snuba.web.db_query.execute_query_with_rate_limits")
+    def test_correct_error_codes(
+        self, execute_mock: MagicMock, record_query: MagicMock
+    ) -> None:
+        # pytest.param doesn't play well with patch so put the list here
+        tests = [
+            (
+                RateLimitExceeded(
+                    "stuff",
+                    scope=TABLE_RATE_LIMIT_NAME,
+                    name="things",
+                ),
+                "table-rate-limited",
+                "against",
+            ),
+            (
+                RateLimitExceeded(
+                    "stuff",
+                    scope=PROJECT_RATE_LIMIT_NAME,
+                    name="things",
+                ),
+                "rate-limited",
+                "for",
+            ),
+            (
+                ClickhouseError("test", code=ErrorCodes.TOO_SLOW),
+                "predicted-timeout",
+                "for",
+            ),
+            (
+                ClickhouseError("test", code=ErrorCodes.TIMEOUT_EXCEEDED),
+                "clickhouse-timeout",
+                "against",
+            ),
+            (
+                ClickhouseError("test", code=ErrorCodes.SOCKET_TIMEOUT),
+                "network-timeout",
+                "against",
+            ),
+            (
+                ClickhouseError("test", code=ErrorCodes.ILLEGAL_TYPE_OF_ARGUMENT),
+                "invalid-typing",
+                "for",
+            ),
+            (
+                ClickhouseError("test", code=ErrorCodes.MEMORY_LIMIT_EXCEEDED),
+                "memory-exceeded",
+                "for",
+            ),
+            (TimeoutError("test"), "cache-set-timeout", "against"),
+            (ExecutionTimeoutError("test"), "cache-wait-timeout", "against"),
+        ]
+
+        for exception, status, slo in tests:
+            execute_mock.side_effect = exception
+            self.post()
```

Comment:
It might seem counter-intuitive to test something so far down the query pipeline with an API call like this, but it was just very cumbersome to try and manually build all the objects that `db_query` uses.